modules/imgproc/src/opencl/resize.cl

```diff
@@ -220,6 +220,78 @@ __kernel void resizeLN(__global const uchar * srcptr, int src_step, int src_offs
     }
 }
 
+#elif defined INTER_LINEAR_EXACT
+
+__kernel void precomputeCoeffs(__read_only image2d_t src, __write_only image2d_t coeffsTex, float inv_fx, float inv_fy) {
+    int dx = get_global_id(0);
+    int dy = get_global_id(1);
+    int sx, sy;
+    float fxx, fyy;
+    short ialpha, ibeta;
+
+    fxx = (float)((dx + 0.5) * inv_fx * 0.5 - 0.5);
+    fyy = (float)((dy + 0.5) * inv_fy * 0.5 - 0.5);
```

Comment:
> `#elif defined INTER_LINEAR_EXACT`

Exact means bit-exact, so we can't use float or double operations which are platform-dependent.

In CPU code there are `softfloat` and `softdouble` types for emulation - they use integers inside.

"Exact" code computes using the integer operations only (and due to specific of GPUs which has lack of integer ALUs it could be slow)

Reply:
I am aware. The current code with pre-computation of coefficients uses a set of non-optimized macros for a minimal fixed-point implementation. It does not work yet and I still need time until i can give it some focus.

```OpenCL
#define FP_MUL(a, b) ((a) * (b) >> INTER_RESIZE_COEF_BITS)
#define FP_DIV(a, b) ((a) << INTER_RESIZE_COEF_BITS / (b))
#define FP_FLOOR(a) ((a) >> INTER_RESIZE_COEF_BITS)
#define FP_FRAC(a) ((a) & (INTER_RESIZE_COEF_SCALE - 1))
#define FP_ROUND(a) FP_FLOOR((a) + INTER_RESIZE_COEF_SCALE / 2)
#define FP_TO_FLOAT(a) ((float)(a) / INTER_RESIZE_COEF_SCALE)
#define FLOAT_TO_FP(a) ((short)((a) * INTER_RESIZE_COEF_SCALE))

__kernel void precomputeCoeffs(__global const uchar* src, int src_step, int src_offset, int src_rows, int src_cols, __global short* coeffsTex, int coeffs_step, int coeffs_offset, int coeffs_rows, int coeffs_cols, float inv_fx, float inv_fy, int dst_cols) {
  int dx = get_global_id(0);
  int dy = get_global_id(1);
  int sx, sy;

  // Convert inv_fx and inv_fy to fixed-point
  short inv_fx_fp = FLOAT_TO_FP(inv_fx);
  short inv_fy_fp = FLOAT_TO_FP(inv_fy);

  // Compute fxx and fyy in fixed-point
  short fxx = FP_MUL(FLOAT_TO_FP(dx), inv_fx_fp);
  sx = FP_FLOOR(fxx);
  fxx = FP_FRAC(fxx);

  if (sx < 0) {
    printf("1. Underflow detected at dx=%d, dy=%d\n", dx, dy);
    fxx = 0, sx = 0;
  }

  if (sx >= src_cols - 1) {
    printf("2. Overflow detected at dx=%d, dy=%d\n", dx, dy);
    fxx = 0, sx = src_cols - 1;
  }

  short fyy = FP_MUL(FLOAT_TO_FP(dy), inv_fy_fp);
  sy = FP_FLOOR(fyy);
  fyy = FP_FRAC(fyy);

  if (sy < 0) {
    printf("3. Underflow detected at dx=%d, dy=%d\n", dx, dy);
    fyy = 0, sy = 0;
  }

  if (sy >= src_rows - 1) {
    printf("4. Overflow detected at dx=%d, dy=%d\n", dx, dy);
    fyy = 0, sy = src_rows - 1;
  }

  // Use convert_short2_sat_rte function to convert float2 to short2
  short2 ialpha = convert_short2_sat_rte((float2)(1.f - FP_TO_FLOAT(fxx), FP_TO_FLOAT(fxx)) * INTER_RESIZE_COEF_SCALE);
  short2 ibeta = convert_short2_sat_rte((float2)(1.f - FP_TO_FLOAT(fyy), FP_TO_FLOAT(fyy)) * INTER_RESIZE_COEF_SCALE);

  // Use vstore2 function to write short2 data to buffer
  vstore2((short2)(sx, ialpha.x), dx * 2, coeffsTex + dy * coeffs_step);
  vstore2((short2)(sy, ibeta.x), dx * 2, coeffsTex + (dy + dst_cols) * coeffs_step);

  // Print out the values for debugging
  printf("dx=%d, dy=%d, sx=%d, sy=%d, fxx=%d, fyy=%d, inv_fx_fp=%d, inv_fy_fp=%d\n", dx, dy, sx, sy, fxx, fyy, inv_fx_fp, inv_fy_fp);
}

__kernel void resizeLN(__global const uchar* src, int src_step, int src_offset, int src_rows, int src_cols, __global short* coeffsTex, int coeffs_step, int coeffs_offset, int coeffs_rows, int coeffs_cols, __global uchar* dst, int dst_step, int dst_offset, int dst_rows, int dst_cols, float inv_fx, float inv_fy) {
    int dx = get_global_id(0);
    int dy = get_global_id(1);

    if (dx < dst_cols && dy < dst_rows)
    {
        __global const short * ialpha = coeffsTex + mad24(dy, (dst_cols << 1), (dx << 1));
        __global const short * ibeta  = coeffsTex + mad24((dst_cols + dy), (dst_cols << 1), (dx << 1));
        short inv_fx_fp = FLOAT_TO_FP(inv_fx);
		short inv_fy_fp = FLOAT_TO_FP(inv_fy);
		int sx = mad_sat(convert_short(dx), inv_fx_fp, convert_short(src_offset));
		int sy = mad_sat(convert_short(dy), inv_fy_fp, convert_short(src_offset));
        int2 isx = convert_int2_sat_rte((float2)(sx, sx + 1) * inv_fx);
        int2 isy = convert_int2_sat_rte((float2)(sy, sy + 1) * inv_fy);

        float4 fsx = convert_float4((int4)(sx, sx + 1, sx, sx + 1)) * inv_fx - convert_float4(isx.xyxy);
        float4 fsy = convert_float4((int4)(sy, sy, sy + 1, sy + 1)) * inv_fy - convert_float4(isy.xyxy);

        short4 alpha = convert_short4_rte(fsx * INTER_RESIZE_COEF_SCALE);
        short4 beta = convert_short4_rte(fsy * INTER_RESIZE_COEF_SCALE);

        int src_index0 = mad24(sy, src_step, mad24(sx, TSIZE, src_offset));
        int src_index1 = mad24(sy + 1, src_step, mad24(sx, TSIZE, src_offset));

        float4 data0 = convert_float4(loadpix(src + src_index0));
        float4 data1 = convert_float4(loadpix(src + src_index0 + TSIZE));
        float4 data2 = convert_float4(loadpix(src + src_index1));
        float4 data3 = convert_float4(loadpix(src + src_index1 + TSIZE));

        float4 fres = mad(data0, (float4)(ialpha[0]), data1 * (float4)(ialpha[1]));
        fres = mad(fres, (float4)(ibeta[0]), mad(data2, (float4)(ialpha[0]), data3 * (float4)(ialpha[1])) * (float4)(ibeta[1]));
        fres = fres * (1.0f / (INTER_RESIZE_COEF_SCALE * INTER_RESIZE_COEF_SCALE));

        uchar res = convert_uchar_sat(fres.x);

        storepix(res, dst + mad24(dy, dst_step, mad24(dx, TSIZE, dst_offset)));

        printf("dx=%d, dy=%d, sx=%d, sy=%d, inv_fx=%f, inv_fy=%f\n", dx, dy, sx, sy, inv_fx, inv_fy);
    }
}
```